staging/src/k8s.io/apiserver/pkg/storage/errors.go

```diff
@@ -96,16 +98,30 @@ func NewInvalidObjError(key, msg string) *StorageError {
 	}
 }
 
+// NewCorruptObjError returns a new StorageError, it represents a corrupt object:
+// a) object data retrieved from the storage failed to transform with the given err.
+// b) the given object failed to decode with the given err
+func NewCorruptObjError(key string, err error) *StorageError {
+	return &StorageError{
+		Code:     ErrCodeCorruptObj,
+		Key:      key,
+		InnerErr: err,
+	}
+}
+
 type StorageError struct {
 	Code               int
 	Key                string
 	ResourceVersion    int64
 	AdditionalErrorMsg string
+	InnerErr           error
 }
 
+func (e *StorageError) Unwrap() error { return e.InnerErr }
+
 func (e *StorageError) Error() string {
-	return fmt.Sprintf("StorageError: %s, Code: %d, Key: %s, ResourceVersion: %d, AdditionalErrorMsg: %s",
-		errCodeToMessage[e.Code], e.Code, e.Key, e.ResourceVersion, e.AdditionalErrorMsg)
+	return fmt.Sprintf("StorageError: %s, Code: %d, Key: %s, ResourceVersion: %d, AdditionalErrorMsg: %s, InnerErr: %v",
+		errCodeToMessage[e.Code], e.Code, e.Key, e.ResourceVersion, e.AdditionalErrorMsg, e.InnerErr)
```

Comment:
that "InnerErr" looks a little non-standard.
```suggestion
	return fmt.Sprintf("storage error %q, code %d, key %q, resource version %d, additional error message %q: %v",
		errCodeToMessage[e.Code], e.Code, e.Key, e.ResourceVersion, e.AdditionalErrorMsg, e.InnerErr)
```
Maybe more like that?

Reply:
it will break any code that introspects the error message, like `strings.Contains(err.Error(), "Code: 1")`, this being a refactor PR, can we do this?
```
"StorageError: %s, Code: %d, Key: %s, ResourceVersion: %d, AdditionalErrorMsg: %s, err: %v"
```

Reply:
sounds good

Reply:
Which case do we have where `AdditionalErrorMsg` and `InnerErr` are set? Can that even happen? It's a little strange to have two messages.

Reply:
@sttts  currently, these two functions make use of `AdditionalErrorMsg`
https://github.com/kubernetes/kubernetes/blob/99cc395e2b9870ab9813e7ea4c0940225e30d05d/staging/src/k8s.io/apiserver/pkg/storage/errors.go#L83-L97

Yes, both should not be used together, as an orthogonal follow-up we can drop `AdditionalErrorMsg`, and keep `InnerErr` only, this is in keeping with conventional go error chaining/handling.

The new function `NewCorruptObjError` uses `InnerErr` only:
```
+func NewCorruptObjError(key string, err error) *StorageError {
+	return &StorageError{
+		Code:     ErrCodeCorruptObj,
+		Key:      key,
+		InnerErr: err,
+	}
+}
```

Reply:
right, `AdditionalErrorMsg: errros.New(AdditionalErrorMsg)`. Why not that?

Reply:
okay, I removed `AdditionalErrorMsg` and just keeping the inner error being introduced in this PR.

a quick search in the `kubernetes` org: https://github.com/search?q=repo%3Akubernetes%2Fkubernetes%20AdditionalErrorMsg&type=code shows no reference to `AdditionalErrorMsg`

similar result with https://grep.app/search?q=AdditionalErrorMsg,  nothing  except for vendoring of this package.

@wojtek-t please let me know if you have any concern with this change